File: packages/core/src/shadow/ContextLoader.ts
import * as fs from 'fs';
import * as path from 'path';

export interface ContextLoaderConfig {
  maxTokenBudget: number;
  cacheEnabled: boolean;
  shadowRoot: string;
  projectRoot: string;
}

export interface LoadedContext {
  filePath: string;
  content: string;
  tokenCount: number;
}

export interface ContextLoadResult {
  contexts: LoadedContext[];
  totalTokens: number;
  budgetRemaining: number;
}

interface CacheEntry {
  content: string;
  tokenCount: number;
  accessedAt: number;
}

const DEFAULT_CONFIG: ContextLoaderConfig = {
  maxTokenBudget: 8000,
  cacheEnabled: true,
  shadowRoot: '.codeflow',
  projectRoot: '.',
};

const CHARS_PER_TOKEN = 4;

export class ContextLoader {
  private readonly config: ContextLoaderConfig;
  private readonly cache: Map<string, CacheEntry> = new Map();
  private readonly maxCacheSize = 100;

  constructor(config: Partial<ContextLoaderConfig> = {}) {
    this.config = { ...DEFAULT_CONFIG, ...config };
  }

  async loadContext(intent: string): Promise<ContextLoadResult> {
    const domainDir = path.resolve(this.config.projectRoot, this.config.shadowRoot, 'domain');
    const intentFiles = await this.scanIntentFiles(domainDir);

    const keywords = this.extractKeywords(intent);
    if (keywords.length === 0) {
      return { contexts: [], totalTokens: 0, budgetRemaining: this.config.maxTokenBudget };
    }

    const scored = await Promise.all(
      intentFiles.map(async (filePath) => {
        const content = await this.readWithCache(filePath);
        const fileKeywords = this.extractKeywords(content);
        const score = this.computeRelevance(keywords, fileKeywords);
        return { filePath, content, score };
      }),
    );

    scored.sort((a, b) => b.score - a.score);

    const contexts: LoadedContext[] = [];
    let totalTokens = 0;

    for (const item of scored) {
      if (item.score <= 0) break;

      const tokenCount = this.estimateTokens(item.content);
      if (totalTokens + tokenCount > this.config.maxTokenBudget) {
        const remaining = this.config.maxTokenBudget - totalTokens;
        if (remaining > 100) {
          const truncated = item.content.slice(0, remaining * CHARS_PER_TOKEN);
          const truncatedTokens = this.estimateTokens(truncated);
          contexts.push({ filePath: item.filePath, content: truncated, tokenCount: truncatedTokens });
          totalTokens += truncatedTokens;
        }
        break;
      }

      contexts.push({ filePath: item.filePath, content: item.content, tokenCount });
      totalTokens += tokenCount;
    }

    return {
      contexts,
      totalTokens,
      budgetRemaining: this.config.maxTokenBudget - totalTokens,
    };
  }

  async loadWithDependencies(filePath: string): Promise<ContextLoadResult> {
    const resolvedPath = path.resolve(filePath);
    const contexts: LoadedContext[] = [];
    let totalTokens = 0;

    const intentDocPath = this.resolveIntentDocPath(resolvedPath);
    const mainContent = await this.readFileOrEmpty(intentDocPath);

    if (mainContent) {
      const tokenCount = this.estimateTokens(mainContent);
      if (totalTokens + tokenCount <= this.config.maxTokenBudget) {
        contexts.push({ filePath: intentDocPath, content: mainContent, tokenCount });
        totalTokens += tokenCount;
      }
    }

    const sourceContent = await this.readFileOrEmpty(resolvedPath);
    if (sourceContent) {
      const imports = this.extractImportPaths(sourceContent);

      for (const importPath of imports) {
        if (totalTokens >= this.config.maxTokenBudget) break;

        const depIntentPath = this.resolveImportIntentPath(importPath, resolvedPath);
        const depContent = await this.readFileOrEmpty(depIntentPath);

        if (depContent) {
          const tokenCount = this.estimateTokens(depContent);
          if (totalTokens + tokenCount <= this.config.maxTokenBudget) {
            contexts.push({ filePath: depIntentPath, content: depContent, tokenCount });
            totalTokens += tokenCount;
          }
        }
      }
    }

    return {
      contexts,
      totalTokens,
      budgetRemaining: this.config.maxTokenBudget - totalTokens,
    };
  }

  clearCache(): void {
    this.cache.clear();
  }

  private estimateTokens(text: string): number {
    return Math.ceil(text.length / CHARS_PER_TOKEN);
  }

  private async readWithCache(filePath: string): Promise<string> {
    if (this.config.cacheEnabled) {
      const cached = this.cache.get(filePath);
      if (cached) {
        cached.accessedAt = Date.now();
        return cached.content;
      }
    }

    const content = await this.readFileOrEmpty(filePath);

    if (this.config.cacheEnabled && content) {
      this.evictIfNeeded();
      this.cache.set(filePath, {
        content,
        tokenCount: this.estimateTokens(content),
        accessedAt: Date.now(),
      });
    }

    return content;
  }

  private evictIfNeeded(): void {
    if (this.cache.size < this.maxCacheSize) return;

    let oldestKey = '';
    let oldestTime = Infinity;

    for (const [key, entry] of this.cache) {
      if (entry.accessedAt < oldestTime) {
        oldestTime = entry.accessedAt;
        oldestKey = key;
      }
    }

    if (oldestKey) {
      this.cache.delete(oldestKey);
    }
  }

  private async readFileOrEmpty(filePath: string): Promise<string> {
    try {
      return await fs.promises.readFile(filePath, 'utf-8');
    } catch {
      return '';
    }
  }

  private async scanIntentFiles(dir: string): Promise<string[]> {
    const files: string[] = [];
    try {
      const entries = await fs.promises.readdir(dir, { withFileTypes: true });
      for (const entry of entries) {
        const fullPath = path.join(dir, entry.name);
        if (entry.isDirectory()) {
          const subFiles = await this.scanIntentFiles(fullPath);
          files.push(...subFiles);
        } else if (entry.name.endsWith('.md') || entry.name.endsWith('.intent.md')) {
          files.push(fullPath);
        }
      }
    } catch {
      // directory does not exist
    }
    return files;
  }

  private extractKeywords(text: string): string[] {
    return text
      .toLowerCase()
      .replace(/[^\w\s\u4e00-\u9fff]/g, ' ')
      .split(/\s+/)
      .filter((t) => t.length > 1);
  }

  private computeRelevance(queryKeywords: string[], docKeywords: string[]): number {
    if (queryKeywords.length === 0 || docKeywords.length === 0) return 0;

    const docSet = new Set(docKeywords);
    let matches = 0;

    for (const kw of queryKeywords) {
      if (docSet.has(kw)) matches++;
    }

    return matches / queryKeywords.length;
  }

  private resolveIntentDocPath(sourceFilePath: string): string {
    const projectRoot = path.resolve(this.config.projectRoot);
    const relativePath = path.relative(projectRoot, sourceFilePath);
    const intentRelative = relativePath.replace(/\.(ts|tsx|js|jsx|go|py)$/, '.intent.md');
    return path.join(projectRoot, this.config.shadowRoot, 'domain', intentRelative);
  }

  private resolveImportIntentPath(importPath: string, fromFile: string): string {
    const projectRoot = path.resolve(this.config.projectRoot);
    let resolvedImport: string;

    if (importPath.startsWith('.')) {
      resolvedImport = path.resolve(path.dirname(fromFile), importPath);
    } else {
      resolvedImport = path.resolve(projectRoot, 'src', importPath);
    }

    const extensions = ['.ts', '.tsx', '.js', '.jsx', '.go', '.py'];
    let ext = path.extname(resolvedImport);
    if (!ext || !extensions.includes(ext)) {
      ext = '.ts';
      resolvedImport = `${resolvedImport}${ext}`;
    }

    const relativePath = path.relative(projectRoot, resolvedImport);
    const intentRelative = relativePath.replace(/\.(ts|tsx|js|jsx|go|py)$/, '.intent.md');
    return path.join(projectRoot, this.config.shadowRoot, 'domain', intentRelative);
  }

  private extractImportPaths(sourceCode: string): string[] {
    const imports: string[] = [];

    const tsImportRegex = /import\s+.*?\s+from\s+['"]([^'"]+)['"]/g;
    let match: RegExpExecArray | null;
    while ((match = tsImportRegex.exec(sourceCode)) !== null) {
      imports.push(match[1]);
    }

    const goImportRegex = /import\s+"([^"]+)"/g;
    while ((match = goImportRegex.exec(sourceCode)) !== null) {
      imports.push(match[1]);
    }

    const pyImportRegex = /(?:from\s+(\S+)\s+import|import\s+(\S+))/g;
    while ((match = pyImportRegex.exec(sourceCode)) !== null) {
      imports.push(match[1] || match[2]);
    }

    return imports.filter(Boolean);
  }
}

File: backend/internal/shadow/context_loader.go
package shadow

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"sync"
)

const charsPerToken = 4

// LoadedContext 已加载的上下文。
type LoadedContext struct {
	FilePath   string `json:"file_path"`
	Content    string `json:"content"`
	TokenCount int    `json:"token_count"`
}

// ContextLoadResult 上下文加载结果。
type ContextLoadResult struct {
	Contexts        []LoadedContext `json:"contexts"`
	TotalTokens     int            `json:"total_tokens"`
	BudgetRemaining int            `json:"budget_remaining"`
}

// ContextLoaderConfig 上下文加载器配置。
type ContextLoaderConfig struct {
	MaxTokenBudget int
	CacheEnabled   bool
	ShadowRoot     string
	ProjectRoot    string
}

type cacheEntry struct {
	content    string
	tokenCount int
	accessedAt int64
}

// ContextLoader 上下文聚焦加载器。
type ContextLoader struct {
	mu           sync.RWMutex
	config       ContextLoaderConfig
	cache        map[string]*cacheEntry
	maxCacheSize int
}

// NewContextLoader 创建上下文加载器实例。
func NewContextLoader(config *ContextLoaderConfig) *ContextLoader {
	cfg := ContextLoaderConfig{
		MaxTokenBudget: 8000,
		CacheEnabled:   true,
		ShadowRoot:     ".codeflow",
		ProjectRoot:    ".",
	}
	if config != nil {
		if config.MaxTokenBudget > 0 {
			cfg.MaxTokenBudget = config.MaxTokenBudget
		}
		if config.ShadowRoot != "" {
			cfg.ShadowRoot = config.ShadowRoot
		}
		if config.ProjectRoot != "" {
			cfg.ProjectRoot = config.ProjectRoot
		}
		cfg.CacheEnabled = config.CacheEnabled
	}
	return &ContextLoader{
		config:       cfg,
		cache:        make(map[string]*cacheEntry),
		maxCacheSize: 100,
	}
}

// LoadContext 根据意图语义搜索 .codeflow 文档，返回 token 预算内的相关文档。
func (l *ContextLoader) LoadContext(intent string) (*ContextLoadResult, error) {
	l.mu.Lock()
	defer l.mu.Unlock()

	domainDir := filepath.Join(l.config.ProjectRoot, l.config.ShadowRoot, "domain")
	intentFiles, err := l.scanIntentFiles(domainDir)
	if err != nil {
		return &ContextLoadResult{
			BudgetRemaining: l.config.MaxTokenBudget,
		}, nil
	}

	keywords := extractKeywords(intent)
	if len(keywords) == 0 {
		return &ContextLoadResult{
			BudgetRemaining: l.config.MaxTokenBudget,
		}, nil
	}

	type scoredFile struct {
		filePath string
		content  string
		score    float64
	}

	var scored []scoredFile
	for _, fp := range intentFiles {
		content, err := l.readWithCache(fp)
		if err != nil || content == "" {
			continue
		}
		fileKw := extractKeywords(content)
		score := l.computeRelevance(keywords, fileKw)
		if score > 0 {
			scored = append(scored, scoredFile{filePath: fp, content: content, score: score})
		}
	}

	// 按分数降序排序
	for i := 0; i < len(scored); i++ {
		for j := i + 1; j < len(scored); j++ {
			if scored[j].score > scored[i].score {
				scored[i], scored[j] = scored[j], scored[i]
			}
		}
	}

	var contexts []LoadedContext
	totalTokens := 0

	for _, item := range scored {
		tokenCount := estimateTokens(item.content)
		if totalTokens+tokenCount > l.config.MaxTokenBudget {
			remaining := l.config.MaxTokenBudget - totalTokens
			if remaining > 100 {
				truncLen := remaining * charsPerToken
				if truncLen > len(item.content) {
					truncLen = len(item.content)
				}
				truncated := item.content[:truncLen]
				truncTokens := estimateTokens(truncated)
				contexts = append(contexts, LoadedContext{
					FilePath:   item.filePath,
					Content:    truncated,
					TokenCount: truncTokens,
				})
				totalTokens += truncTokens
			}
			break
		}

		contexts = append(contexts, LoadedContext{
			FilePath:   item.filePath,
			Content:    item.content,
			TokenCount: tokenCount,
		})
		totalTokens += tokenCount
	}

	return &ContextLoadResult{
		Contexts:        contexts,
		TotalTokens:     totalTokens,
		BudgetRemaining: l.config.MaxTokenBudget - totalTokens,
	}, nil
}

// LoadWithDependencies 加载源文件的意图文档及其依赖的意图文档。
func (l *ContextLoader) LoadWithDependencies(filePath string) (*ContextLoadResult, error) {
	l.mu.Lock()
	defer l.mu.Unlock()

	absPath, err := filepath.Abs(filePath)
	if err != nil {
		absPath = filePath
	}

	var contexts []LoadedContext
	totalTokens := 0

	intentDocPath := l.resolveIntentDocPath(absPath)
	mainContent, _ := l.readFileOrEmpty(intentDocPath)

	if mainContent != "" {
		tokenCount := estimateTokens(mainContent)
		if totalTokens+tokenCount <= l.config.MaxTokenBudget {
			contexts = append(contexts, LoadedContext{
				FilePath:   intentDocPath,
				Content:    mainContent,
				TokenCount: tokenCount,
			})
			totalTokens += tokenCount
		}
	}

	sourceContent, _ := l.readFileOrEmpty(absPath)
	if sourceContent != "" {
		imports := l.extractImportPaths(sourceContent)
		for _, imp := range imports {
			if totalTokens >= l.config.MaxTokenBudget {
				break
			}
			depIntentPath := l.resolveImportIntentPath(imp, absPath)
			depContent, _ := l.readFileOrEmpty(depIntentPath)
			if depContent != "" {
				tokenCount := estimateTokens(depContent)
				if totalTokens+tokenCount <= l.config.MaxTokenBudget {
					contexts = append(contexts, LoadedContext{
						FilePath:   depIntentPath,
						Content:    depContent,
						TokenCount: tokenCount,
					})
					totalTokens += tokenCount
				}
			}
		}
	}

	return &ContextLoadResult{
		Contexts:        contexts,
		TotalTokens:     totalTokens,
		BudgetRemaining: l.config.MaxTokenBudget - totalTokens,
	}, nil
}

// ClearCache 清空缓存。
func (l *ContextLoader) ClearCache() {
	l.mu.Lock()
	defer l.mu.Unlock()
	l.cache = make(map[string]*cacheEntry)
}

func estimateTokens(text string) int {
	n := len(text) / charsPerToken
	if len(text)%charsPerToken != 0 {
		n++
	}
	return n
}

func (l *ContextLoader) readWithCache(filePath string) (string, error) {
	if l.config.CacheEnabled {
		if entry, ok := l.cache[filePath]; ok {
			return entry.content, nil
		}
	}

	content, err := l.readFileOrEmpty(filePath)
	if err != nil {
		return "", err
	}

	if l.config.CacheEnabled && content != "" {
		l.evictIfNeeded()
		l.cache[filePath] = &cacheEntry{
			content:    content,
			tokenCount: estimateTokens(content),
		}
	}

	return content, nil
}

func (l *ContextLoader) evictIfNeeded() {
	if len(l.cache) < l.maxCacheSize {
		return
	}
	// 删除第一个找到的条目（简单 LRU 近似）
	for k := range l.cache {
		delete(l.cache, k)
		break
	}
}

func (l *ContextLoader) readFileOrEmpty(filePath string) (string, error) {
	data, err := os.ReadFile(filePath)
	if err != nil {
		return "", nil
	}
	return string(data), nil
}

func (l *ContextLoader) scanIntentFiles(dir string) ([]string, error) {
	var files []string
	err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return nil // 忽略错误继续扫描
		}
		if info.IsDir() {
			return nil
		}
		if strings.HasSuffix(info.Name(), ".md") {
			files = append(files, path)
		}
		return nil
	})
	return files, err
}

func (l *ContextLoader) computeRelevance(queryKw, docKw []string) float64 {
	if len(queryKw) == 0 || len(docKw) == 0 {
		return 0
	}

	docSet := make(map[string]bool, len(docKw))
	for _, kw := range docKw {
		docSet[kw] = true
	}

	matches := 0
	for _, kw := range queryKw {
		if docSet[kw] {
			matches++
		}
	}

	return float64(matches) / float64(len(queryKw))
}

func (l *ContextLoader) resolveIntentDocPath(sourceFilePath string) string {
	projectRoot, _ := filepath.Abs(l.config.ProjectRoot)
	relPath, err := filepath.Rel(projectRoot, sourceFilePath)
	if err != nil {
		relPath = sourceFilePath
	}

	ext := filepath.Ext(relPath)
	intentRel := strings.TrimSuffix(relPath, ext) + ".intent.md"
	return filepath.Join(projectRoot, l.config.ShadowRoot, "domain", intentRel)
}

func (l *ContextLoader) resolveImportIntentPath(importPath, fromFile string) string {
	projectRoot, _ := filepath.Abs(l.config.ProjectRoot)
	var resolvedImport string

	if strings.HasPrefix(importPath, ".") {
		resolvedImport = filepath.Join(filepath.Dir(fromFile), importPath)
	} else {
		resolvedImport = filepath.Join(projectRoot, "src", importPath)
	}

	ext := filepath.Ext(resolvedImport)
	knownExts := map[string]bool{".ts": true, ".tsx": true, ".js": true, ".jsx": true, ".go": true, ".py": true}
	if !knownExts[ext] {
		resolvedImport = resolvedImport + ".ts"
	}

	relPath, err := filepath.Rel(projectRoot, resolvedImport)
	if err != nil {
		relPath = resolvedImport
	}

	ext = filepath.Ext(relPath)
	intentRel := strings.TrimSuffix(relPath, ext) + ".intent.md"
	return filepath.Join(projectRoot, l.config.ShadowRoot, "domain", intentRel)
}

func (l *ContextLoader) extractImportPaths(sourceCode string) []string {
	var imports []string
	lines := strings.Split(sourceCode, "\n")

	for _, line := range lines {
		trimmed := strings.TrimSpace(line)

		// TypeScript/JavaScript: import ... from '...'
		if strings.Contains(trimmed, "from") && (strings.Contains(trimmed, "'") || strings.Contains(trimmed, "\"")) {
			parts := strings.Split(trimmed, "from")
			if len(parts) >= 2 {
				importPart := strings.TrimSpace(parts[len(parts)-1])
				importPart = strings.Trim(importPart, "'\";")
				if importPart != "" {
					imports = append(imports, importPart)
				}
			}
		}

		// Go: import "..."
		if strings.HasPrefix(trimmed, "import") && strings.Contains(trimmed, "\"") {
			start := strings.Index(trimmed, "\"")
			end := strings.LastIndex(trimmed, "\"")
			if start >= 0 && end > start {
				imp := trimmed[start+1 : end]
				if imp != "" {
					imports = append(imports, imp)
				}
			}
		}
	}

	return imports
}

// String 格式化输出。
func (r *ContextLoadResult) String() string {
	return fmt.Sprintf("ContextLoadResult{contexts=%d, totalTokens=%d, budgetRemaining=%d}",
		len(r.Contexts), r.TotalTokens, r.BudgetRemaining)
}
